src/scripts/external/mastercrafted-api.ts

```diff
@@ -0,0 +1,235 @@
+ * @property {string} [id=null]
+ * @property {string} [name=""]
+ * @property {string} [description=""]
+ * @property {RecipeMasterCrafted[]} [recipes=RecipeMasterCrafted[]]
+ * @property {string[]} [tools=[]]
+ * @property {string} [sound=""]
+ * @property {Object} [ownership={}]
+ * @param {(0|1)} [ingredientsInspection=0] (default 0)
+ * @param {(0|1)} [productInspection=0] (default 0)
+ * @property {string} [img=MASTERCRAFTED_CONST.RECIPE_BOOK.IMG]
+ */
+interface RecipeBookMasterCrafted {
+    id: string;
+    name: string;
+    description: string;
+    recipes?: RecipeMasterCrafted[];
+    tools?: string[];
+    sound?: string;
+    ownership?: Record<string, unknown>;
+    ingredientsInspection?: 0 | 1;
+    productInspection?: 0 | 1;
+    img: string;
+}
+
+/**
+ * @typedef {Object} RecipeMasterCrafted
+ * @property {string} [id=null]
+ * @property {string} [name=""]
+ * @property {string} [description=""]
+ * @property {number} [time=null]
+ * @property {string} [macroName=""]
+ * @property {RecipeBookMasterCrafted} [recipeBook=null]
+ * @property {IngredientMasterCrafted[]} [ingredients=[]]
+ * @property {ProductMasterCrafted[]} [products=[]]
+ * @property {string[]} [tools=[]]
+ * @property {string} [sound=""]
+ * @property {Object} [ownership={}]
+ * @param {(0|1)} [ingredientsInspection=0] (default 0)
+ * @param {(0|1)} [productInspection=0] (default 0)
+ * @property {string} [img=MASTERCRAFTED_CONST.RECIPE_BOOK.IMG]
+ */
+interface RecipeMasterCrafted {
+    id?: string;
+    name?: string;
+    description?: string;
+    time?: number;
+    macroName?: string;
+    recipeBook?: RecipeBookMasterCrafted;
+    ingredients?: IngredientMasterCrafted[];
+    products?: ProductMasterCrafted[];
+    tools?: string[];
+    sound?: string;
+    ownership?: Record<string, unknown>;
+    ingredientsInspection?: 0 | 1;
+    productInspection?: 0 | 1;
+    img?: string;
+}
+
+/**
+ * @typedef {Object} ComponentMasterCrafted
+ * @property {string} [id=undefined]
+ * @property {string} [uuid=undefined]
+ * @property {number} [quantity=undefined]
+ * @property {string} [name=undefined]
+ */
+interface ComponentMasterCrafted {
+    id?: string;
+    uuid?: string;
+    quantity?: number;
+    name?: string;
+}
+
+/**
+ * @typedef {Object} IngredientMasterCrafted
+ * @property {string} id
+ * @property {string} [name=null]
+ * @property {ComponentMasterCrafted[]} [components=[]]
+ * @property {RecipeMasterCrafted} [recipe=null]
+ */
+interface IngredientMasterCrafted {
+    id: string;
+    name?: string;
+    components?: ComponentMasterCrafted[];
+    recipe?: RecipeMasterCrafted;
+}
+
+/**
+ * @typedef {Object} ProductMasterCrafted
+ * @property {string} id
+ * @property {string} name
+ * @property {ComponentMasterCrafted[]} components
+ */
+interface ProductMasterCrafted {
+    id: string;
+    name: string;
+    components: ComponentMasterCrafted[];
+}
+
+
+
+class MasterCraftedAPI extends DefaultRecipeAPI implements RecipeAPI {
```

Comment:
I see where you're going with this.

If we were following the existing conventions in the codebase, rather than _extending_ the Recipe API I would _use_ the Recipe API by making it a dependency of the `MasterCraftedAPI`. I'd introduce a new interface to define the contract for the `MasterCraftedAPI` and rename the default implementation (this class) to `DefaultMasterCraftedAPI`.

However, that aside, I'd like to try something different.

I think this feature should re-use the existing import logic in the `DefaultFabricateAPI`. It will need to be broken out into a new type (`ExchangeAPI` implemented by the `DefaultExchangeAPI`) and re-used (depended upon) by the `DefaultMasterCraftedAPI`. We can map to and from the `FabricateExportModel` and have the `ExchangeAPI` take care of the complexities of mapping this to Fabricate's internal model. 

Something like this:

```mermaid
graph LR
    DefaultFabricateAPI(FabricateAPI) -- Uses(FabricateExportModel) --> DefaultExchangeAPI(ExchangeAPI)
    DefaultMasterCraftedAPI(MasterCraftedAPI) -- Uses(FabricateExportModel) --> DefaultExchangeAPI
```

We may need to add a new data store for the `ExchangeAPI` to track relationships between already imported Recipe Books (and maybe other module's data structures) to prevent the re-import from creating new crafting systems. 
 However, what I really think I'd like to do here is either **ask the user to select the data type in the import dialogue** (e.g. Fabricate, or MasterCrafted) or to do some type detection in the import process where the JSON is being read. that way it is either in the context of an existing crafting system, or a brand new import. We can ask the user to select a type on export too.

Reply:
makes sense...but I would have preferred to try to avoid having you rewrite so much code...just for my use case

Reply:
Every use case is equally as important! Plus, it's largely just moving things around in this refactor - the behaviour and contract of the Fabricate API will change, but not initially and not by much. I'll just be moving the import functions down a level.

If you want to take a stab at this and update the PR I'm fine with that. Or I can work directly on your fork if you have "Allow edits and access to secrets by maintainers" enabled to help out a bit more.

Reply:
But I've written so little at the code level that I would say, do as it comes back to you best.